pkg/lib/filesystem/local-storage.go

```diff
@@ -136,71 +140,137 @@ func saveConfig(ctx context.Context, localRepo local.LocalRepo, kitfile *artifac
 }
 
 func saveKitfileLayers(ctx context.Context, localRepo local.LocalRepo, kitfile *artifact.KitFile, ignore ignore.Paths, opts *SaveModelOptions) (layers []ocispec.Descriptor, diffIDs []digest.Digest, err error) {
+
+	var (
+		eg, egCtx = errgroup.WithContext(ctx)
+	)
+
+	type layerResult struct {
+		layer  ocispec.Descriptor
+		diffID digest.Digest
+		index  int
+	}
+
+	results := make(chan layerResult, 100)
+	currentIndex := 0
+	progress := mpb.New(
+		mpb.WithWidth(60),
+		mpb.WithRefreshRate(150*time.Millisecond),
+	)
+
 	if kitfile.Model != nil {
 		if kitfile.Model.Path != "" && !util.IsModelKitReference(kitfile.Model.Path) {
-			mediaType := mediatype.New(opts.ModelFormat, mediatype.ModelBaseType, opts.LayerFormat, opts.Compression)
-			layer, layerInfo, err := saveContentLayer(ctx, localRepo, kitfile.Model.Path, mediaType, ignore)
-			if err != nil {
-				return nil, nil, err
-			}
-			layers = append(layers, layer)
-			diffIDs = append(diffIDs, digest.FromString(layerInfo.DiffId))
-			kitfile.Model.LayerInfo = layerInfo
+			currentIndexCopy := currentIndex
+			currentIndex++
+			eg.Go(func() error {
+				mediaType := mediatype.New(opts.ModelFormat, mediatype.ModelBaseType, opts.LayerFormat, opts.Compression)
+				layer, layerInfo, err := saveContentLayer(egCtx, localRepo, kitfile.Model.Path, mediaType, ignore, progress)
+				if err != nil {
+					return err
+				}
+				kitfile.Model.LayerInfo = layerInfo
+				results <- layerResult{layer: layer, diffID: digest.FromString(layerInfo.DiffId), index: currentIndexCopy}
+				return nil
+			})
 		}
 		for idx, part := range kitfile.Model.Parts {
-			mediaType := mediatype.New(opts.ModelFormat, mediatype.ModelPartBaseType, opts.LayerFormat, opts.Compression)
-			layer, layerInfo, err := saveContentLayer(ctx, localRepo, part.Path, mediaType, ignore)
-			if err != nil {
-				return nil, nil, err
-			}
-			layers = append(layers, layer)
-			diffIDs = append(diffIDs, digest.FromString(layerInfo.DiffId))
-			kitfile.Model.Parts[idx].LayerInfo = layerInfo
+			currentIndexCopy := currentIndex
+			currentIndex++
+			index := idx
+			path := part.Path
+			eg.Go(func() error {
+				mediaType := mediatype.New(opts.ModelFormat, mediatype.ModelPartBaseType, opts.LayerFormat, opts.Compression)
+				layer, layerInfo, err := saveContentLayer(egCtx, localRepo, path, mediaType, ignore, progress)
+				if err != nil {
+					return err
+				}
+				kitfile.Model.Parts[index].LayerInfo = layerInfo
```

Comment:
This is introducing concurrent writes to shared Kitfile fields without synchronization, causing potential data races and corruption. The same pattern is repeated on other layers too.